[clang][dataflow] Refactor `widen` API to be explicit about change effect.

ymand opened the pull request:
The previous API relied on pointer equality of inputs and outputs to signal
whether a change occured. This was too subtle and led to bugs in practice. It
was also very limiting: the override could not return an equivalent (but not
identical) value.

ymand committed:
[clang][dataflow] Refactor `widen` API to be explicit about change effect.

The previous API relied on pointer equality of inputs and outputs to signal
whether a change occured. This was too subtle and led to bugs in practice. It
was also very limiting: the override could not return an equivalent (but not
identical) value.

martinboehme commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -166,17 +166,20 @@ static Value *joinDistinctValues(QualType Type, Value &Val1,
-                                  const Environment &PrevEnv, Value &Current,
-                                  Environment &CurrentEnv,
-                                  Environment::ValueModel &Model) {
+namespace {
+using WidenResult = Environment::ValueModel::WidenResult;
+}
+
+static WidenResult widenDistinctValues(QualType Type, Value &Prev,
```
```suggestion
static Environment::ValueModel::WidenResult widenDistinctValues(QualType Type, Value &Prev,
```

AFAICT, this is used exactly once. Maybe just spell it out?

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -805,6 +805,25 @@ class NullPointerAnalysis final
+
+  std::optional<WidenResult> widen(QualType Type, Value &Prev,
+                                   const Environment &PrevEnv, Value &Current,
+                                   Environment &CurrentEnv) override {
```
This appears to be more than an API change in that this model didn't have a widening operation at all before. Can you explain why this has been added? (Is it necessary as part of this patch, or just an additional embellishment?)

ymand replied:
I added this so that we can test the widening functionality. Turns out we had no tests for framework calls to `widen`.

martinboehme replied:
Wow -- makes sense to add this then!

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -975,6 +994,35 @@ TEST_F(WideningTest, DistinctValuesWithSamePropertiesAreEquivalent) {
+      Code,
+      [](const llvm::StringMap<DataflowAnalysisState<NoopLattice>> &Results,
+         ASTContext &ASTCtx) {
+        ASSERT_THAT(Results.keys(), UnorderedElementsAre("p"));
```
We do this in a lot of existing tests, but it's unnecessary: `getEnvironmentAtAnnotation()` itself asserts that the annotation exists.

```suggestion
```

ymand replied:
Great. I've removed it here and in all other cases of WideningTest.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -975,6 +994,35 @@ TEST_F(WideningTest, DistinctValuesWithSamePropertiesAreEquivalent) {
+        ASSERT_THAT(Results.keys(), UnorderedElementsAre("p"));
+        const Environment &Env = getEnvironmentAtAnnotation(Results, "p");
+
+        const ValueDecl *FooDecl = findValueDecl(ASTCtx, "Foo");
+        ASSERT_THAT(FooDecl, NotNull());
+
+        const auto *FooVal = Env.getValue(*FooDecl);
```
```suggestion
        const auto &FooVal = getValueForDecl<Value>(ASTCtx, Env, "Foo");
```

ymand replied:
Done

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -975,6 +994,35 @@ TEST_F(WideningTest, DistinctValuesWithSamePropertiesAreEquivalent) {
+        ASSERT_THAT(FooDecl, NotNull());
+
+        const auto *FooVal = Env.getValue(*FooDecl);
+        EXPECT_TRUE(areEquivalentValues(*FooVal->getProperty("is_null"),
```
I think we also want to test that `getProperty()` returns non-null?

Easiest way might be `*cast<BoolValue>(FooVal->getProperty("is_null"))`. (`cast` asserts that its argument is non-null.)

ymand replied:
I prefer direct assertion, because it results in a cleaner error message on failure (crashing the test process vs exiting the test with a test-assertion failure).

martinboehme replied:
Looks good!

martinboehme reviewed (changes requested):
Nice. This is _much_ clearer!

Mostly nits, the only substantive question I have is how we will migrate existing overrides.

ymand commented on clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h:
```diff
@@ -122,17 +135,19 @@ class Environment {
     ///  `PrevEnv` and `CurrentEnv`, respectively.
-    virtual Value *widen(QualType Type, Value &Prev, const Environment &PrevEnv,
-                         Value &Current, Environment &CurrentEnv) {
+    virtual std::optional<WidenResult> widen(QualType Type, Value &Prev,
```
That's the *only* override I could find. So, I think we'll just prepare a patch to crubit to coincide with pushing this.

martinboehme replied:
Makes sense, thanks for the explanation.

ymand committed:
fixup! [clang][dataflow] Refactor `widen` API to be explicit about change effect.

address comments

ymand committed:
fixup! fixup! [clang][dataflow] Refactor `widen` API to be explicit about change effect.

address comments

ymand commented:
Martin, I've addressed all of your comments. PTAL.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -975,6 +989,35 @@ TEST_F(WideningTest, DistinctValuesWithSamePropertiesAreEquivalent) {
+         ASTContext &ASTCtx) {
+        const Environment &Env = getEnvironmentAtAnnotation(Results, "p");
+
+        const ValueDecl *FooDecl = findValueDecl(ASTCtx, "Foo");
+        ASSERT_THAT(FooDecl, NotNull());
+
```
```suggestion
```

`getValueForDecl()` asserts that the declaration exists.

ymand committed:
fixup! [clang][dataflow] Refactor `widen` API to be explicit about change effect.

address comments

martinboehme commented:
> I also fixed up the other tests to use `getValue/LocForDecl` to be consistent.

That's great -- thanks for doing this! I have always been meaning to do a cleanup but have never gotten round to it.

ymand committed:
fixup! [clang][dataflow] Refactor `widen` API to be explicit about change effect.

rename LatticeJoinEffect for use in new API; move WidenResult out of class (to match ComparisonResult)
